drop-transfer/src/manager.rs

```diff
@@ -76,6 +79,27 @@ impl TransferManager {
         self.transfers.get(&id).map(|state| &state.connection)
     }
 
+    pub(crate) fn is_file_cancelled(&self, id: Uuid, file: &FileId) -> crate::Result<bool> {
+        let xstate = self.transfers.get(&id).ok_or(crate::Error::BadTransfer)?;
+        Ok(xstate.cancelled.contains(file))
+    }
+
+    pub(crate) fn cancel_file(&mut self, id: Uuid, file: FileId) -> crate::Result<()> {
+        let xstate = self
+            .transfers
+            .get_mut(&id)
+            .ok_or(crate::Error::BadTransfer)?;
+        if !xstate.xfer.files().contains_key(&file) {
+            return Err(crate::Error::BadFileId);
+        }
+
+        if !xstate.cancelled.insert(file) {
+            return Err(crate::Error::FileCancelled);
+        }
```

Comment:
Since double cancelling doesn't functionally break anything here, and the app teams rely on events always being emitted makes me think, should we return an error here?

Reply:
Good point. IDK. @LukasPukenis what do you think?

Reply:
Both ways seem to be valid approaches :)

We might leave it as it was but provide an additional flag to indicate if the file was in transit when the cancellation happened. This way:
- we always emit the event
- indicate if the file was in flight, so no information hiding is present


I am personally leaning toward what I've described above as we don't break the API needlessly and it sounds simpler. We can also release it as v4.* instead of waiting for v5

@matszczygiel As there's no clear winner in this case I suggest proceeding as you wish

Reply:
@Lipt0nas @LukasPukenis I'm not sure if we all have the same picture.
My understanding is this: the cancel is no longer related to the file being in transit (well it can potentially stop the transit if the file is in flight). Cancel is meant for disabling file transit and it cannot be undone.
So the question is: Since the cancel is idempotent, should we emit an error if the cancel is called more than once?